x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -256,6 +296,13 @@ public void getRoleForApiKey(Authentication authentication, CompositeRolesStore
      */
     static void validateApiKeyCredentials(Map<String, Object> source, ApiKeyCredentials credentials, Clock clock,
                                           ActionListener<AuthenticationResult> listener) {
+        final Boolean invalidated = (Boolean) source.get("api_key_invalidated");
+        if (invalidated == null) {
+            listener.onResponse(AuthenticationResult.terminate("api key document is missing invalidated field", null));
+        } else if (invalidated) {
+            listener.onResponse(AuthenticationResult.terminate("api key has been invalidated", null));
+        }
```

Comment:
you need an else here. Without it the listener could get called with the result from the rest of the method. @jkakavas knows about this :)